### ISSUE TITLE
[no-restricted-imports] add allow type import options

### ISSUE THREAD

**Issue by yoyo930021** [eslint/eslint#13758]
**What rule do you want to change?**
Rule name: no-restricted-imports
Add `allowType` in rule options.

https://github.com/typescript-eslint/typescript-eslint/issues/2661#issuecomment-706570884

**Does this change cause the rule to produce more or fewer warnings?**
Fewer warnings when special case.

**How will the change be implemented? (New option, new default behavior, etc.)?**
Check `importKind` property in `ImportDeclaration`.

**Please provide some example code that this change will affect:**

```js
// wrong
import A from './a'

// correct
import type A from './a'
```

**What does the rule currently do for this code?**
We need to limit only type import some packages sometimes.
For example:
When using package in user project runtimes, Like vetur, ts-loader.

**What will the rule do after it's changed?**
We make sure that only type import is used.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by mdjermanovic** [comment]
Hi @yoyo930021, thanks for the issue! 

> Add `allowType` in rule options.

Non-JS specific options are out of scope for core rules. We do make changes to rules to better work with TypeScript and other language extensions, but I don't think we have any rules or options that aren't applicable to JS code.

You can easily configure `no-restricted-syntax` for this behavior:

```json
{
  "no-restricted-syntax": [ "error", {
    "selector": "ImportDeclaration[importKind!='type'][source.value='./a']",
    "message": "'import type' only from './a'."
  }]
}
```

I'll leave this issue open for the case that other members have a different opinion.

**Comment by bradzacher** [comment]
So my question is if we document this property as an extension to the ESTree spec, would you guys be open to considering it then?

Or is it just a flat "no" thing to non-standard properties?

**Comment by mdjermanovic** [comment]
We do use [estree extensions](https://github.com/estree/estree/tree/master/extensions) specs, for example here:

https://github.com/eslint/eslint/blob/b5e011c865e95d700d29cb9a4ba71c671d99e423/lib/rules/arrow-parens.js#L153-L154

If it makes sense to never restrict `import type`, then we can use the new property to just ignore those imports.

I don't think there's a precedent in core rules for an option that doesn't affect JS code. It would require TypeScript (or some other language) examples in the documentation. I'd say it's out of scope, but would like more opinions.

**Comment by bradzacher** [comment]
Ultimately I'm happy for us to take this as another extension rule in `typescript-eslint`, but now that the ecosystem is more mature I'd love to rely _less_ on extension rules where possible as users need to learn about them before they can use them, leading to users that just don't use them (which means that they have a broken linting experience, which harms the entire ecosystem).

Additionally extension rules live within `@typescript-eslint/eslint-plugin`, so even though this extension would apply to flow, flow users likely wouldn't ever see it.

To reiterate - happy to close this an extend the rule in @typescript-eslint, but I would much rather document this as an estree extension and implement it in here for all to use with zero overhead.

**Comment by mdjermanovic** [comment]
@bradzacher just to clarify, is the proposal for this rule to add an option like `allowType`, or just to always skip type-only imports?

**Comment by bradzacher** [comment]
The former - there are cases where you want to disallow a module to be type or value imported entirely, and there are cases where you want to only disallow a module to be value imported, so having an `allowType` flag would let you differentiate between the two.

I don't think the latter is a good solution as it means we'd have to fork the codebase entirely and build a whole new rule to support the aforementioned usecase.

**Comment by mdjermanovic** [comment]
I'm in favor of using estree extension specs and/or modifying algorithms to work better with language extensions, and we are already doing that wherever possible, but an option specifically for language extensions would be something new in the core rules.

@eslint/eslint-tsc thoughts about this?

**Comment by kaicataldo** [comment]
Currently, adding syntax to ESTree is the best way to get support in ESLint core. 

I personally am always in favor of making `@babel/eslint-parser` and `@typescript-eslint/parser` more first class citizens - because so many people use these parsers - but at this point, we I don't think we have the resources to do this. We're struggling as it is to implement core changes, and new issues are constantly created for the existing core rules.

Core rules maintained by the ESLint team is currently a pain point that gets distributed to every level of the ESLint community for a few reasons:

1. They take up a lot of the resources of the core team to maintain, making it much more difficult for us to improve core itself.
2. Because we don't have the resources to support other common flavors of JS in core rules, it ends up being a huge pain point for parser/plugin maintainers (thinking particularly of Babel and TypeScript), as they have to reimplement core rules and figure out ways to work around the current API which really treats them as second class citizens.
3. The Babel/TypeScript communities have to deal with more complicated configuration/missing rule features that users of "standard" JS get.
4. This translates into more issues/maintenance burden for ESLint core and plugin/parser maintainers, and the cycle repeats.

Given these constraints, I think the ideal outcome would be for the community to eventually take on maintenance of core rules (as a plugin in a separate org), freeing us up to reallocate those resources to core itself. If we ever hit that point, I'd really love to explore how to make Babel and TypeScript have first class support in core, or at least create an API that makes it easier for custom parsers to integrate with rules.

Sorry, this ended up being more of a brain dump of what I've been following for a long time now. Any ideas of how we could improve things would be really appreciated ❤️

**Comment by nzakas** [comment]
I also see this as out of scope. I added the ESTree type definition extensions because we needed some way to just throw away type information in order to work at all with TypeScript and Flow. With type imports, we can't just throw away the `types` keyword because it changes what `import` means, which can have downstream effects on rules and other components.

I think it's safe to say that anything TypeScript-specific belongs in the plugin.

**Comment by bradzacher** [comment]
> I think it's safe to say that anything TypeScript-specific belongs in the plugin.

No problem!
Happy to close this and we'll create an extension rule downstream.

**Comment by nzakas** [comment]
Thanks @bradzacher 🙏 You're doing a great job with the plugin.
